OSDOCS-8015: Adding a total resource usage table (#77127)
openshift/openshift-docs: modules/network-observability-total-resource-usage.adoc

// Module included in the following assemblies:
// * network_observability/configuring_operator.adoc

:_mod-docs-content-type: REFERENCE
[id="network-observability-total-resource-usage-table_{context}"]

> = Resource considerations

Reviewer: ```suggestion
= Total average memory and CPU usage
```

> The following table outlines examples of total resource usage for clusters with a sampling size of 1.

Reviewer: ```suggestion
The following table outlines averages of total resource usage for clusters with sampling sizes of 1, 50, and 400.
```

> The following table outlines averages of total resource usage for clusters with a sampling sizes of 1, 50, and 400.

Reviewer 2: Not sure if "sampling size" is correct , I would tend to say "sampling value" and not consider it as a _size_, wdyt? Same for the table header where it's "Sample size".

An alternative would be "Sampling ratio" and present it as 1:1, 1:50, 1:400

Reviewer: I like sampling value instead of sampling size here. Thanks for the options for different wording :)

Author: Thank you! Updated the headers and the paragraph with the "Sampling value" terminology.

[IMPORTANT]
====
The examples outlined in the table demonstrate scenarios that are tailored to specific workloads. Consider each example only as a baseline from which adjustments can be made to accommodate your workload needs.
====

> .Total resource usage

Reviewer: ```suggestion
.Total average resource usage
```

[options="header"]
|===

> | Sample size | Parameters | Small (25 nodes)  | Medium (65 nodes) ^[2]^ | Large (120 nodes) ^[2]^

Reviewer 2: I think we need to use different headers for the columns. The main driver is not the size of the cluster, but it's the test being run. So it should be perhaps something like _Test 1_, _Test 2_, _Test 3_, with the cluster size in parenthesis. I'm insisting on this because otherwise, the numbers sometimes don't make sense. For instance, with sampling 400 we see that netobserv uses more CPU than on a medium cluster. That sounds wrong, but it's just that the test scenario is completely different.

Reviewer 2: cc @memodi

Author: Updated the headers to _Test 1, Test 2, Test 3_ along with the corresponding cluster sizes in parentheses.

Reviewer 3: @jotak @mpandya15 - I am fine with having as Test 1, 2, 3 headers , I thinking that may be too abstract though, do we need to provide some visibility as to what each _Test_ is about?

Reviewer: I agree @memodi . Its my feeling that Test 1, 2, and 3 and S, M and L are both generic naming conventions. But i don't think we can say node-density-heavy, ingress-perf, and cluster-density-v2. Would a customer have any idea what those are? I would guess no, but not sure if there's an alternative way to categorize these.

Reviewer: I agree @memodi . Its my feeling that Test 1, 2, and 3 and S, M and L are both generic naming conventions. But i don't think we can say node-density-heavy, ingress-perf, and cluster-density-v2. Would a customer have any idea what those are? I would guess no, but not sure if there's an alternative way to categorize these.

Reviewer: Whatever we decide here wrt naming conventions, we will need to apply it to the existing Resource consideration table updates as well.

Reviewer 2: or we could add a footnote on each test, with a short description:

- Test 1 focuses on pods density per node, each of them generating a low traffic, on a small size cluster
- Test 2 focuses on ingress traffic throughput, on a medium size cluster
- Test 3 also focuses on pods density (like Test 1), running on larger cluster and generating more traffic

Reviewer 2: cc @stleerh in case you want to chime in here

Reviewer 3: short description as suggested [here](https://github.com/openshift/openshift-docs/pull/77127#discussion_r1657146887) sounds good to me. If we want to give specific metrics that represents scale (as OCP perf doc does) we collect `namespaceCount/podStatusCount/serviceCount` metrics for node/cluster-density workloads. For ingress-perf, we probably would need to rely on a metric that represents throughput of bytes (`nWorkloadBytesProcessedPerMinuteCadvisor`?).

Author: Hi @jotak, thanks for your input, I've updated the footnotes. In response to @memodi's comment here about adding metrics in footnotes, these would be good details to add but more certainty would be needed around the `ingress-perf` metrics. Just needed to clarify whether this is something you recommend adding to the existing docs or a detail for future reference? Thank you! 😊

.6+| *Sampling = 1* | *With Loki* 3+|
| Total NetObserv CPU Usage | 3.24 | 3.42 | 7.30
| Total NetObserv RSS (Memory) Usage | 14.09 GB | 11.81 GB | 36.46 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 2.40 | 2.43 | 5.59
| Total NetObserv RSS (Memory) Usage | 6.85 GB | 10.39 GB | 13.92 GB
.6+| *Sampling = 50* | *With Loki* 3+|
| Total NetObserv CPU Usage | 2.04 | 2.36 | 3.31
| Total NetObserv RSS (Memory) Usage | 8.79 GB | 19.14 GB | 21.07 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 1.55 | 1.64 | 2.70
| Total NetObserv RSS (Memory) Usage | 6.71 GB | 10.15 GB | 14.82 GB
.6+| *Sampling = 400* | *With Loki* 3+|
| Total NetObserv CPU Usage | 1.71 | 1.44 | 2.36
| Total NetObserv RSS (Memory) Usage | 8.21 GB | 16.02 GB | 17.44 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 1.31 | 1.06 | 1.83
| Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB
|===
[.small]
--
1. Tested with AWS M6i instances.
2. In addition to this worker and its controller, 3 infra nodes (size `M6i.12xlarge`) and 1 workload node (size `M6i.8xlarge`) were tested.
--

> Summary: Total resource usage of NetObserv (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.

Reviewer: ```suggestion
Summary: Average total resource usage of Network Observability (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.
```

> Summary: Average total resource usage of Network Observability (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.

Reviewer 2: This sentence doesn't correspond to the numbers presented above. We can either adapt it, or completely remove it?

> Summary: Average total resource usage of Network Observability (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.

Author: @jotak @memodi Thanks for the clarification! Since the table is more comprehensive than just sampling value of 1, the summary now provides a generic statement that this table only presents the average total resource usage of Network Observability (Agents+FLP+Kafka+Loki). 

CC: @skrthomas :)

// Module included in the following assemblies:
// * network_observability/configuring_operator.adoc

:_mod-docs-content-type: REFERENCE
[id="network-observability-total-resource-usage-table_{context}"]
= Total average memory and CPU usage

> The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400.

Reviewer: ```suggestion
The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400 for 3 different tests: `Test 1`, `Test 2`, and `Test 3`. The tests differ in the following ways:
- Test 1 and 3 take into account the `namespaceCount`, `podStatusCount` and `serviceCount` metrics , and represents use cases with a high number of workloads.
- Test 2 takes into account a high ingress traffic throughput.

Since different types of cluster use cases are exemplified in the different tests, these numbers cannot be linearly compared side-by-side but instead used as a benchmark to evaluate what your personal cluster usage may be.
```

Reviewer: Based on [Slack](https://redhat-internal.slack.com/archives/C02939DP5L5/p1720812777670899?thread_ts=1720793630.658669&cid=C02939DP5L5) thread, we need to add more context about whats in the tests as a kind of "expectation-setting" to introduce the table.

Reviewer: cc @stleerh @memodi wdyt about adding some expectation-setting to the introduction, like this? Anything you would add here?

Reviewer 3: @skrthomas thanks, overall it looks good to me. Instead of listing each prometheus metric names that we defined in our tests, I'd say `... take into account total number of Namespace, Pods and Services in OCP cluster ... with a high number of workloads for a given cluster size`

Reviewer: Ok, cool. That makes sense, thank you!

Reviewer 2: That looks good to me, thanks @skrthomas 
I think given the tests that we run, this is well formulated. But still, this is IMO not what users would expect when they look for reference values. I think for 1.7 we should run consistent tests for different cluster sizes (like NDH-cni) for this doc - even if we continue to run other tests internally for perf regression tracking.

Reviewer: @jotak  so if I understand, this is as good as we can give with the tests that we run today? And there's future potential to run consistent tests for different cluster sizes?

Reviewer 2: Yes exactly, I hope in 1.7 we can simplify the doc by showing data with the same test used for each cluster size

Reviewer: Do you think we should wait to publish this until 1.7, and wait on those updates? Or should we add these now?

====

.Total average resource usage
[%autowidth, options="header"]
|===
| Sampling value | Parameters | Test 1 (25 nodes) ^[1]^ | Test 2 (65 nodes) ^[2]^ | Test 3 (120 nodes) ^[3]^
.6+| *Sampling = 1* | *With Loki* 3+|
| Total NetObserv CPU Usage | 3.24 | 3.42 | 7.30
| Total NetObserv RSS (Memory) Usage | 14.09 GB | 11.81 GB | 36.46 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 2.40 | 2.43 | 5.59
| Total NetObserv RSS (Memory) Usage | 6.85 GB | 10.39 GB | 13.92 GB
.6+| *Sampling = 50* | *With Loki* 3+|
| Total NetObserv CPU Usage | 2.04 | 2.36 | 3.31
| Total NetObserv RSS (Memory) Usage | 8.79 GB | 19.14 GB | 21.07 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 1.55 | 1.64 | 2.70
| Total NetObserv RSS (Memory) Usage | 6.71 GB | 10.15 GB | 14.82 GB
.6+| *Sampling = 400* | *With Loki* 3+|
| Total NetObserv CPU Usage | 1.71 | 1.44 | 2.36
| Total NetObserv RSS (Memory) Usage | 8.21 GB | 16.02 GB | 17.44 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 1.31 | 1.06 | 1.83

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer: Mehul said one decimal place is fine, so we can go to one.

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer: So when Manushri went to round these, there's a couple of times where this creates too similar looking data. For example, *Without Loki*
| Total NetObserv CPU Usage | 2.4 | 2.4 | 5.6

I think for this reason, we should stick to 2 decimal places. 

cc @mpandya15

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer 4: Maybe we could add `total % CPU usage` and `total % memory usage` in seperate rows or inside the same cells ? 

That was a question that often came discussing with customers.

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer 4: Also, should we mention prometheus impact here ? cc @jotak

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer: > Maybe we could add `total % CPU usage` and `total % memory usage` in seperate rows or inside the same cells ?
> 
> That was a question that often came discussing with customers.

@jpinsonneau I was chatting w/ @memodi about this, and he says we don't have these metrics now, but it will be something we can capture in future release.

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer 4: Can't we calculate the totals from the nodes hardware specs ? :thinking: 
Then it's a simple rule of three

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer 3: @jpinsonneau I don't know if we want total Netobserv % CPU/RSS Usage as a function of available on nodes OR as a function of cluster usage - from discussions with @stleerh we need latter (which I think makes sense) for which new prometheus queries needs to be added. Also, if we look w.r.t  nodes hardware it would be minuscule usage since that would be dependent on number of nodes in a cluster we obviously have several in each of our tests.

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer: For the Prometheus impact, if @jotak agrees, we can add a note like:
[IMPORTANT]
====
Metrics exported to Prometheus can impact the resource usage, so its important to be aware of the cardinality of the values you're exporting. For more information, see Flows format reference. 
====

Then link to the Flows format reference

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer 2: @skrthomas sgtm!

> | Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB

Reviewer 5: @memodi @jpinsonneau It should be with respect to cluster usage.  If it's based on availability, you can allocate a lot of resources without using it so that the percent used by NetObserv is low.

|===

--

> 1. Test 1 focuses on pods density per node, each node generating low traffic on a small size cluster.

Reviewer 3: ```suggestion
1. Test 1 focuses on pods density per node, each pod generating low amounts of traffic on a small size cluster.
```

2. Test 2 focuses on ingress traffic throughput on a medium size cluster.

> 3. Test 3 focuses on pods density, generating more traffic on a large size cluster.

Reviewer 3: are these suppose to be numbered bullets or of citation style?

Reviewer: @memodi The 1, 2, and 3 under the table are the footnotes for [1], [2], and [3] for each Test X column heading. Is that what you mean?

Reviewer: Previously, we used to be able to use a `[.small]` attribute in asciidoc to make the footnotes look small, like actual footnote size. But (doc tools woes 😢 ) the Customer Portal docs don't render the `[.small]` attribute. Even though docs.openshift.com renders it fine, our contributing guide states not to use it because it isn't fully supported in all the places we publish. Not sure if you're wondering why the numbers look so huge, like code callouts.

Reviewer 3: yes, that's what I was referring too, I thought it needs to be same of style [1], [2] or [3] as they were citied for e.g. `Test 1 (25 nodes) ^[1]^`

> 3. Test 3 focuses on pods density, generating more traffic on a large size cluster.

Reviewer 3: ```suggestion
3. Test 3 focuses on pods density, generating heavy traffic on a large size cluster.
```

> 3. Test 3 focuses on pods density, generating heavy traffic on a large size cluster.

Reviewer 2: in fact we shouldn't say "heavy traffic" for Test 3, it's not heavy. I think at first my suggestion was to just say "more traffic than in Test 1"

--

> Summary: This table shows average total resource usage of Network Observability (Agents+FLP+Kafka+Loki).

Reviewer: I think that "With" and "Without" are clearer here for accessibility considerations (thinking of screen readers).

// Module included in the following assemblies:
// * network_observability/configuring_operator.adoc

:_mod-docs-content-type: REFERENCE
[id="network-observability-total-resource-usage-table_{context}"]
= Total average memory and CPU usage

The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400 for 3 different tests: `Test 1`, `Test 2`, and `Test 3`. The tests differ in the following ways:

> - `Test 1` takes into account the total number of namespace, pods and services in {product-title} cluster, places load on the eBPF agent, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 1` consists of 76 Namespaces, 5153 Pods, and 2305 Services.

Reviewer 6: ```suggestion
- `Test 1` takes into account the total number of namespace, pods and services in an {product-title} cluster, places load on the eBPF agent, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 1` consists of 76 Namespaces, 5153 Pods, and 2305 Services.
```

> - `Test 2` takes into account a high ingress traffic throughput.

Reviewer 6: Could "throughput" be replaced with "capacity" ? Not sure, but tbh had to look up definition of that word. :)

Reviewer 2: then maybe "bandwidth" rather than "capacity", would that work for you?

Reviewer 6: Hello @jotak,
   Thanks for the response. I think either would work...and to be honest, "throughput" was not a term that I was familiar with...maybe others would be, just trying to make it as clear as possible. Thanks!

Author: Switched it up with "volume." Thanks for the suggestion. :)

> - `Test 3` takes into account the total number of namespace, pods and services in {product-title} cluster, places load on the eBPF agent on a larger scale than `Test 1`, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 3` consists of 553 Namespaces, 6998 Pods, and 2508 Services.

Reviewer 6: ```suggestion
- `Test 3` takes into account the total number of namespace, pods and services in an {product-title} cluster, places load on the eBPF agent on a larger scale than `Test 1`, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 3` consists of 553 Namespaces, 6998 Pods, and 2508 Services.
```

Since different types of cluster use cases are exemplified in the different tests, the numbers in this table cannot be linearly compared side-by-side, but instead are intended to be used as a benchmark for evaluating your personal cluster usage.

[NOTE]
====

> Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted.. For more information, see Flows format reference.

Reviewer 6: ```suggestion
Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted. For more information, see Flows format reference.
```

> Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted.. For more information, see Flows format reference.

Reviewer 6: ```suggestion
Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted.. For more information, see "Network Flows format" in the Additional resources section.
```

> ====

Reviewer: Since we have these two admonitions back-to-back, which I know I am the one who said we should put these things in admonitions, maybe we can work one of them into the main text. Like I think the IMPORTANT can maybe be incorporated into the main text, then leave the NOTE. So that last paragraph would be like this:
> Since different types of cluster use cases are exemplified in the different tests, the numbers in this table cannot be linearly compared side-by-side, but instead are intended to be used as a benchmark for evaluating your personal cluster usage. The examples outlined in the following table demonstrate scenarios that are tailored to specific workloads. Consider each example only as a baseline from which adjustments can be made to accommodate your workload needs. 

> [NOTE]
====
Metrics exported to Prometheus.....
====

Reviewer: Here's the SSG reference for not stacking admonitions: https://redhat-documentation.github.io/supplementary-style-guide/#admonitions

[IMPORTANT]
====
The examples outlined in the table demonstrate scenarios that are tailored to specific workloads. Consider each example only as a baseline from which adjustments can be made to accommodate your workload needs.
====

> .Total average resource usage

Reviewer 6: Maybe reference each test in the table? In other words, instead of just providing sampling value as indicator of which test is which, instead identify test 1, 2 and 3?

| Total NetObserv RSS (Memory) Usage | 14.09 GB | 22.56 GB | 36.46 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 2.40 | 2.43 | 5.59
| Total NetObserv RSS (Memory) Usage | 6.85 GB | 10.39 GB | 13.92 GB
.6+| *Sampling = 50* | *With Loki* 3+|
| Total NetObserv CPU Usage | 2.04 | 2.36 | 3.31
| Total NetObserv RSS (Memory) Usage | 8.79 GB | 19.14 GB | 21.07 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 1.55 | 1.64 | 2.70
| Total NetObserv RSS (Memory) Usage | 6.71 GB | 10.15 GB | 14.82 GB
.6+| *Sampling = 400* | *With Loki* 3+|
| Total NetObserv CPU Usage | 1.71 | 1.44 | 2.36
| Total NetObserv RSS (Memory) Usage | 8.21 GB | 16.02 GB | 17.44 GB
| *Without Loki* 3+|
| Total NetObserv CPU Usage | 1.31 | 1.06 | 1.83
| Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB
|===

--
1. Test 1 focuses on pods density per node, each pod generating low amounts of traffic on a small size cluster.
2. Test 2 focuses on ingress traffic throughput on a medium size cluster.
3. Test 3 focuses on pods density, generating more traffic than Test 1 on a large size cluster.

> --

Reviewer 3: @mpandya15 I think we can remove this now that you have provided detailed info about these tests before the table ?
